File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.py
def solution(bridge_length, weight, truck_weights):
    answer = 1
    bridge=[[truck_weights.pop(0),0]]
    while bridge or truck_weights:
        answer+=1
        w=0
        if bridge:
            for i in range(len(bridge)):
                w+=bridge[i][0]
                bridge[i][1]+=1
            if bridge[0][1]==bridge_length-1:
                bridge.pop(0)
        if truck_weights:
            if w+truck_weights[0]<=weight and len(bridge)<=bridge_length:
                bridge.append([truck_weights.pop(0),0])

    return answer+1

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.java
import java.util.*;

class Solution {

    public int solution(int bridge_length, int weight, int[] truck_weights) {

        Queue<Integer> bridge = new LinkedList<>();

        int time = 0;
        int curWeight = 0;
        int idx = 0;

        // 다리 초기 상태
        for (int i = 0; i < bridge_length; i++) {
            bridge.offer(0);
        }

        while (!bridge.isEmpty()) {

            time++;

            // 한 칸 이동
            curWeight -= bridge.poll();

            // 다음 트럭 올릴 수 있으면
            if (idx < truck_weights.length &&
                curWeight + truck_weights[idx] <= weight) {

                bridge.offer(truck_weights[idx]);
                curWeight += truck_weights[idx];
                idx++;
            }
            else {
                bridge.offer(0);
            }

            // 모든 트럭 끝났으면 종료
            if (idx == truck_weights.length && curWeight == 0) {
                break;
            }
        }

        return time;
    }
}